File: linear_regression.py
import numpy as np
import random

# Seed for reproducibility
seed = 42
np.random.seed(seed)
random.seed(seed)

# Generate dataset
X = np.array([2 * np.random.random() for _ in range(100)])
y = 4 + 3 * X + np.random.normal(0, 1, size=X.shape)

# Split into training and test sets
train_size = int(X.size * 0.8)
X_train = X[:train_size]
y_train = y[:train_size]
X_test = X[train_size:]
y_test = y[train_size:]

print(f"trainSize: {train_size}")
print(f"X_train.size: {X_train.size}; X_train: {X_train}")
print(f"X_test.size: {X_test.size}; X_test: {X_test}")
print(f"y_train.size: {y_train.size}; y_train: {y_train}")
print(f"y_test.size: {y_test.size}; y_test: {y_test}")

# Add bias term to X_train and X_test
X_train_b = np.c_[np.ones(X_train.shape[0]), X_train]
X_test_b = np.c_[np.ones(X_test.shape[0]), X_test]

# Initialize weights
theta = np.random.randn(2)

# Learning rate
learning_rate = 0.01
n_iterations = 1000
m = X_train.size

# Gradient Descent
for iteration in range(n_iterations):
    y_pred = X_train_b.dot(theta)
    errors = y_train - y_pred

    gradients = -2.0 / m * X_train_b.T.dot(errors)
    theta -= learning_rate * gradients

# Make predictions on the test set
y_test_pred = X_test_b.dot(theta)

# Calculate Mean Squared Error (MSE)
mse = np.mean((y_test - y_test_pred) ** 2)

print(f"Theta: {theta}")
print(f"Mean Squared Error (MSE): {mse}")

File: linear_regression.kt
import java.util.Random

fun main() {
    val seed = 42L
    val random = Random(seed)

    // Generate dataset
    val X = (1..100).map { 2 * random.nextDouble() }
    val y = X.map { 4 + 3 * it + random.nextGaussian() } // Using Gaussian for noise

    // Split into training and test sets
    val trainSize = (X.size * 0.8).toInt()
    val X_train = X.take(trainSize)
    val y_train = y.take(trainSize)
    val X_test = X.drop(trainSize)
    val y_test = y.drop(trainSize)

    // Print dataset sizes
    println("trainSize: $trainSize")
    println("X_train.size: ${X_train.size}; X_train: $X_train")
    println("X_test.size: ${X_test.size}; X_test: $X_test")
    println("y_train.size: ${y_train.size}; y_train: $y_train")
    println("y_test.size: ${y_test.size}; y_test: $y_test")

    // Add bias term to X_train and X_test
    val X_train_b = X_train.map { listOf(1.0, it) }
    val X_test_b = X_test.map { listOf(1.0, it) }

    // Initialize weights
    var theta = listOf(random.nextGaussian(), random.nextGaussian())

    // Learning rate
    val learningRate = 0.01
    val nIterations = 1000
    val m = X_train.size

    // Gradient Descent
    for (iteration in 0 until nIterations) {
        val yPred = X_train_b.map { it[0] * theta[0] + it[1] * theta[1] }
        val errors = y_train.zip(yPred) { yi, ypi -> yi - ypi }

        val gradients = List(2) { j ->
            -2.0 / m * errors.zip(X_train_b) { ei, xi -> ei * xi[j] }.sum()
        }

        theta = theta.zip(gradients) { t, g -> t - learningRate * g }
    }

    // Make predictions on the test set
    val y_trainPred = X_train_b.map { it[0] * theta[0] + it[1] * theta[1] }
    val yTestPred = X_test_b.map { it[0] * theta[0] + it[1] * theta[1] }

    // Calculate Mean Squared Error (MSE)
    val mse = y_test.zip(yTestPred) { yi, ypi -> (yi - ypi).pow(2) }.average()

    println("Theta: $theta")
    println("Mean Squared Error (MSE): $mse")


}

// Extension function for power
fun Double.pow(exp: Int): Double {
    return Math.pow(this, exp.toDouble())
}
